lightning/src/ln/channelmanager.rs

```diff
@@ -1528,29 +1529,48 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 	/// Note that ALL inputs in the transaction pointed to by funding_txo MUST spend SegWit outputs
 	/// or your counterparty can steal your funds!
 	///
+	/// Returns an [`APIError::APIMisuseError`] if the funding_transaction spent non-SegWit outputs
+	/// or the output didn't match the parameters in [`Event::FundingGenerationReady`].
+	///
 	/// Panics if a funding transaction has already been provided for this channel.
 	///
 	/// May panic if the funding_txo is duplicative with some other channel (note that this should
 	/// be trivially prevented by using unique funding transaction keys per-channel).
-	pub fn funding_transaction_generated(&self, temporary_channel_id: &[u8; 32], funding_txo: OutPoint) {
+	pub fn funding_transaction_generated(&self, temporary_channel_id: &[u8; 32], funding_transaction: Transaction, output_index: u16) -> Result<(), APIError> {
 		let _persistence_guard = PersistenceNotifierGuard::new(&self.total_consistency_lock, &self.persistence_notifier);
 
+		for inp in funding_transaction.input.iter() {
+			if inp.witness.is_empty() {
+				return Err(APIError::APIMisuseError {
+					err: "Funding transaction must be fully signed and spend Segwit outputs".to_owned()
+				});
+			}
+		}
+		let funding_txo = OutPoint { txid: funding_transaction.txid(), index: output_index };
+
 		let (chan, msg) = {
 			let (res, chan) = match self.channel_state.lock().unwrap().by_id.remove(temporary_channel_id) {
 				Some(mut chan) => {
-					(chan.get_outbound_funding_created(funding_txo, &self.logger)
+					if output_index as usize >= funding_transaction.output.len() ||
+							funding_transaction.output[output_index as usize].script_pubkey != chan.get_funding_redeemscript().to_v0_p2wsh() ||
+							funding_transaction.output[output_index as usize].value != chan.get_value_satoshis() {
+						return Err(APIError::APIMisuseError {
+							err: "Funding Transaction Output did not match the script_pubkey or value in the FundingGenerationReady event".to_owned()
+						});
+					}
+					(chan.get_outbound_funding_created(funding_transaction, funding_txo, &self.logger)
 						.map_err(|e| if let ChannelError::Close(msg) = e {
 							MsgHandleErrInternal::from_finish_shutdown(msg, chan.channel_id(), chan.force_shutdown(true), None)
 						} else { unreachable!(); })
 					, chan)
 				},
-				None => return
+				None => { return Err(APIError::ChannelUnavailable { err: "No such channel".to_owned() }) },
 			};
 			match handle_error!(self, res, chan.get_counterparty_node_id()) {
 				Ok(funding_msg) => {
 					(chan, funding_msg)
 				},
-				Err(_) => { return; }
+				Err(_) => { return Err(APIError::ChannelUnavailable { err: "Error deriving keys - either our RNG or our counterparty's RNG is broken".to_owned() }) },
```

Comment:
Hmm, couldn't it have been another error? Seems that we can reach this point only if `sign_counterparty_commitment` fails

Reply:
Hmm, I was looking at the `build_remote_transaction_keys()` failures, though you're right, `sign_counterparty_commitment` is a user function, so it could be for any reason (including hardware wallet disconnected). If its our own `InMemorySigner`, I think that's still the only case.